### DISCUSSION TITLE
iOS status

### DISCUSSION THREAD

**Discussion by jdevng** [swift-quic/swift-quic#17] [Q&A]
What is the status of quic for iOS - I noticed the missing check mark. I have a python client and server that uses quic, and would like to connect with iOS. Is it close to working, or did it need work from the ground up?

**Comment by btoms20** [comment]
Hey! If you're making http3 requests, iOS 15+ should [support it using NetworkManager already](https://developer.apple.com/documentation/technotes/tn3102-http3-in-your-app). The only requirement is that your server has to be running draft-29 or above of QUIC.

If, on the other hand, you're trying to open / manipulate raw QUIC connections / streams then you might need to wait for this project, or a similar one, to offer it. 

We're still a ways away from having an http3/QUIC client ready for swift-nio apps. But it's on the radar and once we get one working/developed, it should run on iOS out of the box. Unfortunately I can't offer you a timeline though. 

Let me know if that helped answer your question or if I misinterpreted what you were asking.

**Reply by jdevng** [reply]
I've been using that example, which brought me to this project. I'm a hardware network engineer, and developing both a software version of the server and a chip version.

Yes, I'd like to go down to the low level UDP traffic for RFC9000 - but this is for a prototype, so not expecting anything to work or exist. The way I'm viewing this is as a avionics chip engineer - I believe draft 29 is from about a year earlier, I'm working from May 2021.

Really what I have an issue with isn't the technical aspects or writing low level software, it's understanding how current implemented software is partitioned, what features are supported, etc. It's the domain specific language that i can't determine how it maps to RFC9000.

I'd be glad to drive a branch development (My software experience may not be the level you require organizationally or syntax-wise to be a part of this development) but what I'd like to do is long term create the 22 frames with connections and streams as soon as possible for iOS/OSX. I'm expecting my effort to be in compliance with RFC9000 May 2021, in terms of wireshark, I wouldn't expect my code to be worthy. I could test that implementation against my software and silicon, also in simulation, and it could serve as a 'ice breaker' or reference for future development or evolution.

If that development proposal is acceptable/interesting, I need a few things to get started on this right away. The first thing would be a single someone motivated to assist me in getting this out - the assistance isn't the technical side. I'd need a text based PoC to asynchronously answer my sideband questions on code structure/domain language issues as map the RFC to swift.

**Reply by btoms20** [reply]
Gotchya! 

We're more than happy for contributions to be made to this repo! 

I would first suggest checking out the [dev branch](https://github.com/swift-quic/swift-quic/tree/develop) and seeing the progress made so far. We're currently able to establish connections both as a server and a client, handle muxed streams and rotate keys. What we're mainly lacking at the moment is loss detection, congestion control and a user focused API for connection / stream management.

You can target your python server by repurposing the [External Go Client Test here](https://github.com/swift-quic/swift-quic/blob/develop/Tests/QuicTests/NIOTests/GoInteropTests/QuicClientGoServerHandshakeTests.swift). This test should handshake with your server, establish a connection, and allow you to start playing around with opening streams and sending data. 

> **Note** 
> Because we don't support http3 yet, you'll have to adjust the ALPNs that your server supports and the ALPN (applicationProtocols) the client uses in that test to match, otherwise the handshake won't succeed. You can see we use a custom `quic-echo-example` ALPN for testing purposes, but this can be anything, as long as the client / sever match.
